File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0785.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        
        #define BLACK 0
        #define WHITE 1
        vector<int> color(n, WHITE);
        
        #define SETA 0
        #define SETB 1
        vector<int> set(n);  // Which set does the node belong to? 
        
        // Since the graph may be disconnected, we may have to start a new BFS
        // for each connected component. This is why we have this for loop
        queue<int> Q;
        for (int i=0; i<n; i++){
            if (color[i] == WHITE){
                // Start a BFS at node "i"
                color[i] = BLACK;
                set[i] = SETA;
                Q.push(i);
                
                while (!Q.empty()){
                    int u = Q.front(); Q.pop();
                    
                    for (int j=0; j<graph[u].size(); j++){
                        int v = graph[u][j];
                        
                        if (color[v] == BLACK){
                            if (set[u] == set[v]) return false;
                        } else {
                            color[v] = BLACK;
                            Q.push(v);
                            set[v] = set[u] == SETA? SETB: SETA;
                        } // end-else
                    } // end-for
                } // end-while
                
            } // end-if
        } // end-for
        
        return true;
    } // end-isBipartite
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0785.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        
        BLACK = 0
        WHITE = 1
        color = [WHITE]*n
        
        SETA = 0
        SETB = 1
        set = [SETA]*n
        
        # Since the graph may be disconnected, we may have to start a new BFS
        # for each connected component. This is why we have this for loop
        Q = []
        for i in range(n):
            if (color[i] == WHITE):
                # Start a BFS at node "i"
                color[i] = BLACK
                set[i] = SETA
                Q.append(i)
                
                while (len(Q) > 0):
                    u = Q.pop()
                    
                    for j in (range(len(graph[u]))):
                        v = graph[u][j]
                        
                        if (color[v] == BLACK):
                            if (set[u] == set[v]): return False
                        else:
                            color[v] = BLACK
                            set[v] = SETB if set[u] == SETA else SETB
                            Q.append(v)
        
        return True
